core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala

```diff
@@ -180,6 +180,56 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
    * as in scala.TraversableOnce. The former operation is used for merging values within a
    * partition, and the latter is used for merging values between partitions. To avoid memory
    * allocation, both of these functions are allowed to modify and return their first argument
+   * instead of creating a new U. This method is different from the ordinary "aggregateByKey"
+   * method, it directly returns a map to the driver, rather than a rdd. This will also perform
+   * the merging locally on each mapper before sending results to a reducer, similarly to a
```

Comment:
doesn't `aggregateByKey` perform map side combine?

Reply:
Yeah, it will. Here the 'difference' means it directly returns a map to the driver rather than an RDD.

Reply:
what's the difference between `aggregateByKeyLocally` and `aggregateByKey(...).toLocalIterator`?

Reply:
`aggregateByKey(...).toLocalIterator` need a shuffle for `aggregateByKey` and then collect the `RDD` to driver as a iterator. But `aggregateByKeyLocally` seems like the `aggregateByKey`, while there isn't a shuffle. It calculates the combines in each task and then collect all the `map` direcly to driver and do the finally combines on driver.

Reply:
> collect all the `map` directly to driver

technically it's a shuffle too, and generally `aggregateByKey` is better for the following reasons:
1. The final combine is executed on multiple reducers, which has better parallelism than doing it on the driver.
2. We should always prefer doing computation on executors instead of the driver, because the driver is responsible for scheduling and has a high cost of failure recovery.
3. using spark shuffle is better for fault tolerant. If one reducer failed, you don't need to rerun all mappers.

So I'm -1 on this API. If there are special cases we wanna to local aggregate, just call `RDD.map.collect` and do the local aggregate manually.

Reply:
BTW `collect all the map directly to driver` may easily OOM the driver, while shuffling to multiple reducers can reduce the memory pressure. Even if only shuffle to one reducer, it still better as the executor usually have more memory than the driver.

Reply:
thansk for the advance, I'll close it and try `mapPartitions(...).collect` in `NaiveBayes`.